[Push Notifications fix] Few improvements

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10327 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR has three main changes:
1. Added a tracks event to track the order notification sound, the sound will be tracked in the event `NEW_ORDER_PUSH_NOTIFICATION_SOUND`, and will have a property `sound_status` that reflects the status of the sound.
2. Updated the logic of fixing notification sound to support the case where the sound was modified and not disabled.
3. Added tracking for the notification fixing solution.

**Some context regarding 2:**
When we implemented the first solution, we were focusing on accidental notification changes (more context here p1701328346130389-slack-C03L1NF1EA3), but since we are getting many complaints about the usage of the app's default sound, I updated the solution to support this case too. I added the logic to the settings entry, the dialog we show from the Orders list will still be shown only when sound is disabled, I think that showing it for anyone who changed the sound could be intrusive, but please let me know what you think.

### Testing instructions
##### TC 1: Tracking
1. Open the app.
2. Check logcat for an entry: `Tracked: new_order_push_notification_sound, Properties: {"sound_status": { DEFAULT | DISABLED | SOUND_MODIFIED }`

##### TC 2: Fix disabled sound from orders list
1. Open the notifications menu in App Information on your device.
2. Open the "New order alerts" entry, and disable its sound.
3. Open the app.
4. Navigate to orders list.
5. Notice the alert that explains the issue.
6. Confirm an event `new_order_push_notification_fix_shown` was tracked.
7. Tap on "Enable sound"
8. Confirm a Snackbar was shown.
9. Confirm an event `new_order_push_notification_fix_tapped` with `source = order_list` was tracked.

##### TC 3: Fix disabled sound from settings
1. Open the notifications menu in App Information on your device.
2. Open the "New order alerts" entry, and disable its sound.
3. Open the app.
4. Open app settings, then tap on Notifications.
5. Confirm the existence of a button "Enable Cha-Ching sound"
6. Tap on the button.
8. Confirm a Snackbar was shown.
9. Confirm an event `new_order_push_notification_fix_tapped` with `source = settings` was tracked.

##### TC 4: Fix modified sound from settings
1. Open the notifications menu in App Information on your device.
2. Open the "New order alerts" entry.
3. Select a different notification sound.
4. Repeat 3 - 9 from TC 3.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Bring back the notification channel log message

hichamboushaba committed:
Add explicit notification channel initialization

This is useful to have the a consistent logging message on app start

hichamboushaba committed:
Add logic to track new order notif sound when channel already exists

hichamboushaba committed:
Check for importance when checking for sound status

hichamboushaba committed:
Add ability to restore order notification sound when modified

hichamboushaba committed:
Add analytics for the order notification sound fix

hichamboushaba committed:
Update unit tests

hichamboushaba committed:
Fix detekt issue

hichamboushaba committed:
Throw an exception when tracking an event before initialization

This will help identify potential bugs during the development

hichamboushaba committed:
Add a release notes entry

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/199">17.8</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/analytics/AnalyticsTracker.kt:
```diff
@@ -680,6 +680,9 @@ class AnalyticsTracker private constructor(
         fun track(stat: AnalyticsEvent, properties: Map<String, *> = emptyMap<String, String>()) {
+            if (instance == null && BuildConfig.DEBUG) {
+                error("event $stat was tracked before AnalyticsTracker was initialized.")
+            }
```
I added this to warn developers about this potential issue, when I first added the notification tracking, I accidentally added it before the call to initialize Analytics, and I didn't notice the issue until I was checking for the event in Logcat.

hichamboushaba committed:
Disable tracks initialization check on tests

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/analytics/AnalyticsTracker.kt:
```diff
@@ -680,6 +681,9 @@ class AnalyticsTracker private constructor(
         }
 
         fun track(stat: AnalyticsEvent, properties: Map<String, *> = emptyMap<String, String>()) {
+            if (instance == null && BuildConfig.DEBUG && !PackageUtils.isTesting()) {
+                error("event $stat was tracked before AnalyticsTracker was initialized.")
+            }
```
Good idea! 👍🏼

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/NotificationChannelsHandler.kt:
```diff
@@ -86,4 +104,17 @@ class NotificationChannelsHandler @Inject constructor(
         )
     }
+
+    private fun NotificationChannel.getNewOrderNotificationSoundStatus(): NewOrderNotificationSoundStatus {
+        if (importance < NotificationManager.IMPORTANCE_DEFAULT) return NewOrderNotificationSoundStatus.DISABLED
+        return when (sound) {
+            context.getChaChingUri() -> NewOrderNotificationSoundStatus.DEFAULT
+            null -> NewOrderNotificationSoundStatus.DISABLED
+            else -> NewOrderNotificationSoundStatus.SOUND_MODIFIED
+        }
+    }
```
Hmm, I think this is not working as expected. I'm seeing the following event upon starting the app: `Tracked: new_order_push_notification_sound, Properties: {"sound_status":"SOUND_MODIFIED"`. However, I have the `chaChing` sound enabled. 

When I debug this code I can see: 
`sound` = `android.resource://com.woocommerce.android.dev/2131886080`
`context.getChaChingUri()` = `android.resource://com.woocommerce.android.dev/2131951616`

Unsure why the `R.raw.cha_ching` is different

hichamboushaba replied:
Oh, really, weird, I'll relax the condition a bit, and you can test it later.

hichamboushaba replied:
Can you test with the last commit 6357d2f4fd982d4ba6e62b2ec398bde360e700d6?

JorgeMucientes replied:
Re tested! Works as expected now! Thanks for the quick fix 🙇🏼

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -764,17 +765,25 @@ class OrderListViewModel @Inject constructor(
+                            AnalyticsEvent.NEW_ORDER_PUSH_NOTIFICATION_FIX_TAPPED,
+                            mapOf(AnalyticsTracker.KEY_SOURCE to "order_list")
+                        )
+                        recreateNotificationChannel()
```
Minor, but the snackbar we are displaying upoin tapping this button is too fast. I didn't have enough time to click on `TEST SOUND`. I think it would make sense to increase the duration of it.

hichamboushaba replied:
The action snackbars all over the app [use](https://github.com/woocommerce/woocommerce-android/blob/df35ff5da8bb66c83130db084b7427acf5bfd158/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/UIMessageResolver.kt#L318) the default long duration, if you are OK with this, let's keep it as is, and if needed we can start a discussion internally about the duration for all of them, because the `Undo` snackbars are more important than this, and if the duration is short, then we need to increase it for them too.

JorgeMucientes replied:
Yes, that's true. Had not noticed that the SnackBar `event` doesn't accept setting the duration. I think the duration also should depend on the length of the message. We need to provide enough time for the user to read, and then think if they want to click on the CTA. 
For `UNDO` actions the current time might be enough, as there's clear context on what is going on. 
In any case this is a really minor thing and the refactor is bigger than I though, since we cannot pass a custom duration to the `SnackBar` event. So, its fine to leave it as it is.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -764,17 +765,25 @@ class OrderListViewModel @Inject constructor(
         }
 
-        if (!notificationChannelsHandler.checkNotificationChannelSound(NotificationChannelType.NEW_ORDER) &&
+        if (notificationChannelsHandler.checkNewOrderNotificationSound() == NewOrderNotificationSoundStatus.DISABLED &&
```
Minor but with this logic, it will behave differently from Notification > Settings. Because here, we are not showing the explanatory dialog when the Notification sound has been changed only when its disabled, so user won't be able to restore the ChaChing sound from here.

hichamboushaba replied:
Yes, that was on purpose as I explained in the PR description, the disabling of the sound can happen accidentally (p1701347291432849/1701328346.130389-slack-C03L1NF1EA3) so we use this kind of "intrusive" dialog to let the user know about it, but the sound change shouldn't be (the tracks will help us confirm this), so we just add the Settings entry to allow the user to restore it without any disturb to their flow, WDYT?

JorgeMucientes replied:
True! I missed that. I got confused by following the last part of the testing instructions and missed that part of the descriptions. The current behavior makes sense. Sorry for the noise.

JorgeMucientes commented:
Excellent job @hichamboushaba. I left some minor comments that it would be nice to address. However, I think its worth merging this asap, so it makes it in time for the freeze today. 
Let me know your thoughts so we can give green light to @samiuelson to start the freeze as soon as this is merged, or if he can proceed without merging this. PR.

hichamboushaba committed:
Relax the condition for the cha-ching sound check

It seems the resource ID could change from app update to another (maybe after changing resources)

hichamboushaba committed:
Merge branch 'trunk' into issue/10327-push-notifications-sound-improvements

hichamboushaba committed:
Suppress detekt issue
